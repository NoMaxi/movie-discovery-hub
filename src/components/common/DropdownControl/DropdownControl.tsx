import { useRef, useState } from "react";
import { useClickOutside } from "@/hooks/useClickOutside/useClickOutside";
import SelectArrow from "@/components/common/SelectArrow/SelectArrow";
import { ContextMenu } from "@/components/common/ContextMenu/ContextMenu";

export interface DropdownControlProps<T extends string> {
    label: string;
    options: readonly T[];
    currentSelection: T;
    onSelectionChange: (newSelection: T) => void;
    testId?: string;
    className?: string;
    buttonClassName?: string;
    dropdownClassName?: string;
}

export const DropdownControl = <T extends string>({
    label,
    options,
    currentSelection,
    onSelectionChange,
    testId,
    className = "",
    buttonClassName = "",
    dropdownClassName = "",
}: DropdownControlProps<T>) => {
    const [isOpen, setIsOpen] = useState(false);
    const controlRef = useRef<HTMLDivElement>(null);

    useClickOutside(controlRef, () => setIsOpen(false), isOpen);

    const toggleDropdown = () => {
        setIsOpen(!isOpen);
    };

    const closeMenu = () => {
        setIsOpen(false);
    };

    const handleOptionSelect = (option: T) => {
        onSelectionChange(option);
        closeMenu();
    };

    const contextMenuActions = options.map((option) => ({
        label: option,
        onClick: () => handleOptionSelect(option),
    }));

    return (
        <div
            className={`dropdown-control relative inline-block text-left ${className}`}
            ref={controlRef}
            data-testid={testId}
        >
            <div className="flex items-center gap-x-4 text-base">
                <span className="dropdown-control-label text-[16px] uppercase opacity-60">{label}</span>
                <button
                    type="button"
                    className={`
                        dropdown-control-button inline-flex items-center justify-center gap-x-2 min-w-[140px]
                        text-[16px] font-medium uppercase
                        focus:outline-none cursor-pointer
                        ${buttonClassName}
                    `}
                    onClick={toggleDropdown}
                >
                    {currentSelection}
                    <span className={`transform transition-transform ${isOpen ? "rotate-180" : "rotate-0"}`}>
                        <SelectArrow />
                    </span>
                </button>
            </div>

            <ContextMenu
                isOpen={isOpen}
                onClose={closeMenu}
                actions={contextMenuActions}
                className={`
                    dropdown-control-dropdown
                    absolute right-0
                    mt-2 w-44 origin-top-right rounded
                    ${dropdownClassName}
                `}
            />
        </div>
    );
};
