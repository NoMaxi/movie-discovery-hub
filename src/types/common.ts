export type Genre = "All" | "Documentary" | "Comedy" | "Horror" | "Crime";

export interface Movie {
    id: string;
    imageUrl: string;
    title: string;
    releaseYear: number;
    genres: string[];
}
