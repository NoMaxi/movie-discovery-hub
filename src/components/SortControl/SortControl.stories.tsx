import type { Meta, StoryObj } from "@storybook/react";
import { fn } from "@storybook/test";
import { SortOption } from "@/types/common";
import SortControl from "./SortControl";

const SORT_OPTIONS: SortOption[] = ["Release Date", "Title"];

const meta = {
    component: SortControl,
    argTypes: {
        currentSelection: {
            control: "select",
            options: SORT_OPTIONS,
            description: "Currently selected sorting option",
        },
        onSelectionChange: {
            description: "Callback function triggered on sort option selection",
        },
    },
    args: {
        onSelectionChange: fn(),
    },
} satisfies Meta<typeof SortControl>;

export default meta;

type Story = StoryObj<typeof SortControl>;

export const ReleaseDateSelected: Story = {
    args: {
        currentSelection: "Release Date",
    },
};

export const TitleSelected: Story = {
    args: {
        currentSelection: "Title",
    },
};

export const OpenedDropdown: Story = {
    args: {
        currentSelection: "Release Date",
    },
    parameters: {
        componentState: {
            isOpen: true,
        },
    },
};
