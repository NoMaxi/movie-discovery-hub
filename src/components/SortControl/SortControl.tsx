import { useState, useRef } from "react";
import { SortOption } from "@/types/common";
import { useClickOutside } from "@/hooks/useClickOutside/useClickOutside";
import SelectArrow from "@/components/common/SelectArrow/SelectArrow";
import { ContextMenu } from "@/components/common/ContextMenu/ContextMenu";

interface SortControlProps {
    currentSelection: SortOption;
    onSelectionChange: (newSelection: SortOption) => void;
}

export const SortControl = ({ currentSelection, onSelectionChange }: SortControlProps) => {
    const [isOpen, setIsOpen] = useState(false);
    const controlRef = useRef<HTMLDivElement>(null);
    const options: SortOption[] = ["Release Date", "Title"];

    useClickOutside(controlRef, () => setIsOpen(false), isOpen);

    const toggleDropdown = () => {
        setIsOpen(!isOpen);
    };

    const closeMenu = () => {
        setIsOpen(false);
    };

    const contextMenuActions = options.map((option) => ({
        label: option,
        onClick: () => onSelectionChange(option),
    }));

    return (
        <div className="sort-control relative inline-block text-left" ref={controlRef}>
            <div className="flex items-center gap-x-4 text-base">
                <span className="sort-control-label text-[16px] uppercase opacity-60">Sort by</span>
                <button
                    type="button"
                    className="
                        sort-control-button inline-flex items-center justify-center gap-x-2
                        text-[16px] font-medium uppercase
                        focus:outline-none cursor-pointer
                    "
                    onClick={toggleDropdown}
                >
                    {currentSelection}
                    <span className={`transform transition-transform ${isOpen ? "rotate-180" : "rotate-0"}`}>
                        <SelectArrow />
                    </span>
                </button>
            </div>

            <ContextMenu
                isOpen={isOpen}
                onClose={closeMenu}
                actions={contextMenuActions}
                className="
                    sort-control-dropdown
                    absolute right-0
                    mt-2 w-48 origin-top-right rounded
                "
            />
        </div>
    );
};
