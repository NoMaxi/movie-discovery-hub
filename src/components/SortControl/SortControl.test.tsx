import { render, screen } from "@testing-library/react";
import userEvent, { UserEvent } from "@testing-library/user-event";
import "@testing-library/jest-dom";
import { SortOption } from "@/types/common";
import { SortControl } from "./SortControl";
import { DropdownControlProps } from "@/components/common/DropdownControl/DropdownControl";

jest.mock("@/components/common/DropdownControl/DropdownControl", () => ({
    DropdownControl: <T extends string>({
        label,
        options,
        currentSelection,
        onSelectionChange,
        testId,
        className,
    }: DropdownControlProps<T>) => (
        <div data-testid={testId} className={className}>
            <span data-testid="dropdown-label">{label}</span>
            <button
                data-testid="dropdown-button"
                onClick={() => {
                    const currentIndex = options.indexOf(currentSelection);
                    const nextIndex = (currentIndex + 1) % options.length;
                    onSelectionChange(options[nextIndex]);
                }}
            >
                {currentSelection}
            </button>
        </div>
    ),
}));

describe("SortControl", () => {
    let onSelectionChangeMock: jest.Mock;
    let user: UserEvent;

    beforeEach(() => {
        onSelectionChangeMock = jest.fn();
        user = userEvent.setup();
    });

    const renderComponent = (currentSelection: SortOption) => {
        return render(<SortControl currentSelection={currentSelection} onSelectionChange={onSelectionChangeMock} />);
    };

    test("Should render DropdownControl with correct props", () => {
        const currentSelection = "Release Date";
        const { asFragment } = renderComponent(currentSelection);

        expect(screen.getByTestId("sort-control")).toBeInTheDocument();
        expect(screen.getByTestId("sort-control")).toHaveClass("sort-control");
        expect(screen.getByTestId("dropdown-label")).toHaveTextContent("Sort by");
        expect(screen.getByTestId("dropdown-button")).toHaveTextContent(currentSelection);
        expect(asFragment()).toMatchSnapshot();
    });

    test("Should pass correct options to DropdownControl", () => {
        renderComponent("Title");

        expect(screen.getByTestId("dropdown-button")).toHaveTextContent("Title");
    });

    test("Should call onSelectionChange when DropdownControl triggers change", async () => {
        renderComponent("Release Date");
        const button = screen.getByTestId("dropdown-button");

        await user.click(button);

        expect(onSelectionChangeMock).toHaveBeenCalledTimes(1);
        expect(onSelectionChangeMock).toHaveBeenCalledWith("Title");
    });

    test("Should update displayed selection when currentSelection prop changes", () => {
        const { rerender } = renderComponent("Release Date");

        expect(screen.getByTestId("dropdown-button")).toHaveTextContent("Release Date");

        rerender(<SortControl currentSelection="Title" onSelectionChange={onSelectionChangeMock} />);

        expect(screen.getByTestId("dropdown-button")).toHaveTextContent("Title");
    });

    test("Should render with correct testId and className", () => {
        renderComponent("Release Date");

        const sortControl = screen.getByTestId("sort-control");
        expect(sortControl).toBeInTheDocument();
        expect(sortControl).toHaveClass("sort-control");
    });
});
