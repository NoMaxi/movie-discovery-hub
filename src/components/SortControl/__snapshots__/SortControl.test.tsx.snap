// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SortControl Should render correctly with initial selection 1`] = `
<DocumentFragment>
  <div
    class="sort-control relative inline-block text-left"
  >
    <div
      class="flex items-center gap-x-4 text-base"
    >
      <span
        class="sort-control-label text-[16px] uppercase opacity-60"
      >
        Sort by
      </span>
      <button
        class="
                        sort-control-button inline-flex items-center justify-center gap-x-2
                        text-[16px] font-medium uppercase
                        focus:outline-none cursor-pointer
                    "
        type="button"
      >
        Release Date
        <span
          class="transform transition-transform rotate-0"
        >
          <svg
            data-testid="select-arrow"
          />
        </span>
      </button>
    </div>
  </div>
</DocumentFragment>
`;
